pylint/testutils/configuration_test.py

```diff
@@ -0,0 +1,101 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+"""Utility function for configuration testing."""
+import copy
+import json
+import logging
+import unittest
+from pathlib import Path
+from typing import Any, Dict, Tuple, Union
+from unittest.mock import Mock
+
+from pylint.lint import Run
+
+USER_SPECIFIC = str(Path(__file__).parent.parent.parent)
+
+
+def get_expected_or_default(pyproject_toml_path: str, suffix: str, default: Any) -> str:
+    def get_path_according_to_suffix() -> Path:
+        path = Path(pyproject_toml_path)
+        return path.parent / f"{path.stem}.{suffix}"
+
+    expected = default
+    expected_result_path = get_path_according_to_suffix()
+    if expected_result_path.exists():
+        with open(expected_result_path, encoding="utf8") as f:
+            expected = f.read()
+        logging.info("%s exists.", expected_result_path)
+    else:
+        logging.info("%s not found, using '%s'.", expected_result_path, default)
+    return expected
+
+
+EXPECTED_CONF_APPEND_KEY = "functional_append"
+EXPECTED_CONF_REMOVE_KEY = "functional_remove"
+EXPECTED_CONF_SPECIAL_KEYS = [EXPECTED_CONF_APPEND_KEY, EXPECTED_CONF_REMOVE_KEY]
+
+
+def get_expected_configuration(
+    configuration_path: str, default_configuration: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get the expected parsed configuration of a configuration functional test"""
+    result = copy.deepcopy(default_configuration)
+    config_as_json = get_expected_or_default(
+        configuration_path, suffix="result.json", default="{}"
+    )
+    to_override = json.loads(config_as_json)
+    for key, value in to_override.items():
+        if key not in EXPECTED_CONF_SPECIAL_KEYS:
+            result[key] = value
+    if EXPECTED_CONF_APPEND_KEY in to_override:
+        for key, value in to_override[EXPECTED_CONF_APPEND_KEY].items():
+            result[key] += value
+    if EXPECTED_CONF_REMOVE_KEY in to_override:
+        for key, value in to_override[EXPECTED_CONF_REMOVE_KEY].items():
+            new_value = []
+            for old_value in result[key]:
+                if old_value not in value:
+                    new_value.append(old_value)
+            result[key] = new_value
+    return result
+
+
+def get_expected_output(configuration_path: str) -> Tuple[int, str]:
+    """Get the expected output of a functional test."""
+
+    def get_relative_path(path: str) -> str:
+        """Get the relative path we want without user specific information"""
+        return path.replace(USER_SPECIFIC, "")[1:]
+
+    output = get_expected_or_default(configuration_path, suffix="out", default="")
+    if output:
+        logging.info(
```

Comment:
Similar to other comment: do we normally use `logging` in functional tests?

Reply:
Although duplicate I think a comment about why we're using `logging` here will help avoid future confusion.

Reply:
da4c8feb419a54544f076258bf4116bd8cd21fd3